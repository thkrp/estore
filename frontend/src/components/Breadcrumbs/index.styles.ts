import styled from 'styled-components';
import { theme } from '../../styles/theme';
import Link from '../Links/Link';

export const WrapperStyled = styled.div`
    height: 52px;
    background: ${theme.colors.main.background};
`;

export const ListStyled = styled.div`
    height: 100%;
    display: flex;
    align-items: center;
    padding: 0;
`;

export const ElementStyled = styled.div`
    display: flex;
    text-overflow: ellipsis;
    text-decoration: none;
    color: ${theme.colors.darkgrey};
    padding: 0 20px;
    height: 100%;
    position: relative;
    align-items: center;
    font-size: 14px;

    &::after {
        content: '';
        display: inline-block;
        height: 100%;
        width: 1px;
        background: ${theme.colors.borders.grey};
        position: absolute;
        right: 0;
    }

    &::before {
        content: '';
        display: block;
        position: absolute;
        right: -5px;
        width: 10px;
        height: 10px;
        border-right: 1px solid ${theme.colors.borders.grey};
        border-top: 1px solid ${theme.colors.borders.grey};
        z-index: 1;
        transform: rotate(45deg);
        background: ${theme.colors.main.background};
    }

    &:last-child {
        color: ${theme.colors.main.primary};

        &::before,
        &::after {
            display: none;
        }
    }
`;

export const MoreCategoriesButtonStyled = styled.div`
    border: 8px solid transparent;
    border-right: 8px solid ${theme.colors.borders.grey};
    border-bottom: 8px solid ${theme.colors.borders.grey};
    position: absolute;
    right: 0;
    bottom: 0;
    cursor: pointer;

    &:hover {
        border-right-color: ${theme.colors.darkgrey};
        border-bottom-color: ${theme.colors.darkgrey};
    }
`;

export const CategoriesStyled = styled.div`
    position: absolute;
    background: ${theme.colors.borders.lightgrey};
    top: 100%;
    left: 100%;
    z-index: 1;
    padding: 5px;

    a {
        display: block;
        width: 100%;
        white-space: nowrap;
        color: ${theme.colors.darkgrey};
    }
`;

export const LinkStyled = styled(Link)`
    color: inherit;
    text-decoration: none;
`;
