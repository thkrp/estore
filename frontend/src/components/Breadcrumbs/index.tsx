import React, { useEffect, useState } from 'react';
import { ReactComponent as LoaderIcon } from '../../assets/icons/horizontal-loader.svg';
import {
    CategoriesStyled,
    ElementStyled,
    LinkStyled,
    ListStyled,
    MoreCategoriesButtonStyled,
    WrapperStyled
} from './index.styles';
import useBreadcrumbs from '../../common/hooks/use.breadcrumbs';
import { transformToKey } from '../../common/helpers/key.helper';

const Breadcrumbs = () => {
    const breadcrumbs = useBreadcrumbs();
    const [showCategories, setShowCategories] = useState<Record<string, boolean>>({});

    useEffect(() => {
        setShowCategories({});
    }, [breadcrumbs]);

    return (
        <WrapperStyled>
            <ListStyled role="navigation" aria-label="breadcrumbs">
                {breadcrumbs.length
                    ? breadcrumbs.map((el, index) =>
                          index === breadcrumbs.length - 1 ? (
                              <ElementStyled key={transformToKey(index, el.code)}>
                                  {el.name || <LoaderIcon height="15px" role="img" />}
                              </ElementStyled>
                          ) : (
                              <ElementStyled key={transformToKey(index, el.code)}>
                                  <LinkStyled to={el.url}>{el.name || el.code}</LinkStyled>
                                  {el.children && (
                                      <>
                                          {breadcrumbs.length - 1 !== index && (
                                              <MoreCategoriesButtonStyled
                                                  role="presentation"
                                                  onClick={() =>
                                                      setShowCategories(prev => ({
                                                          [el.code]: !prev[el.code]
                                                      }))
                                                  }
                                              />
                                          )}
                                          {showCategories[el.code] && (
                                              <CategoriesStyled role="menu" aria-label="sub-categories">
                                                  {el.children.map((cat, i) => (
                                                      <LinkStyled
                                                          to={cat.url}
                                                          key={transformToKey(i, `${cat.url}`)}
                                                          role="menuitem"
                                                      >
                                                          {cat.name || cat.code}
                                                      </LinkStyled>
                                                  ))}
                                              </CategoriesStyled>
                                          )}
                                      </>
                                  )}
                              </ElementStyled>
                          )
                      )
                    : null}
            </ListStyled>
        </WrapperStyled>
    );
};

export default Breadcrumbs;
