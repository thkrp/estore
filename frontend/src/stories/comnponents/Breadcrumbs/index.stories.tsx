import React from 'react';
import { Meta, StoryObj } from '@storybook/react';
import { mockData } from 'app-shared';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { Breadcrumbs as BreadcrumbsComponent } from '../../../components';

const {
    catalogMenu,
    products: {
        items: [detailed]
    }
} = mockData;
const initialState = {
    app: {
        menu: {
            catalog: catalogMenu
        }
    },
    catalog: {
        detailed
    }
};
const path = '/catalog/motoblochnoe-navesnoe-oborudovanie/adaptery-dlya-motoblokov';
const id = 'adapter-k-motobloku-ara';
const mockStore = configureStore();
let store;
const meta = {
    title: 'Components/Breadcrumbs',
    component: BreadcrumbsComponent,
    parameters: {
        layout: 'centered'
    },
    tags: ['autodocs']
} satisfies Meta<typeof BreadcrumbsComponent>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Categories: Story = {
    decorators: [
        Story => {
            store = mockStore(initialState);
            return (
                <MemoryRouter initialEntries={[`${path}`]}>
                    <Provider store={store}>{Story()}</Provider>
                </MemoryRouter>
            );
        }
    ]
};

export const ProductIsLoading: Story = {
    decorators: [
        Story => {
            store = mockStore({ ...initialState, catalog: { detailed: { name: undefined, code: '', url: '' } } });
            return (
                <MemoryRouter initialEntries={[`${path}/${id}`]}>
                    <Provider store={store}>
                        <Routes>
                            <Route path={`${path}/:id`} element={Story()} />
                        </Routes>
                    </Provider>
                </MemoryRouter>
            );
        }
    ]
};

export const ProductLoaded: Story = {
    decorators: [
        Story => {
            store = mockStore({
                ...initialState,
                catalog: { detailed: { name: 'Адаптер к мотоблоку "АРА"', code: id, url: path } }
            });
            return (
                <MemoryRouter initialEntries={[`${path}/${id}`]}>
                    <Provider store={store}>
                        <Routes>
                            <Route path={`${path}/:id`} element={Story()} />
                        </Routes>
                    </Provider>
                </MemoryRouter>
            );
        }
    ]
};
