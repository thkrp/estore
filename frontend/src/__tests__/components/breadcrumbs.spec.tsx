/* eslint-disable no-restricted-imports*/
import React from 'react';
import '@testing-library/jest-dom';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { mockData } from 'app-shared';
import Breadcrumbs from '../../components/Breadcrumbs';

describe('breadcrumbs', () => {
    afterEach(cleanup);

    const {
        catalogMenu,
        products: {
            items: [detailed]
        }
    } = mockData;
    const initialState = {
        app: {
            menu: {
                catalog: catalogMenu
            }
        },
        catalog: {
            detailed
        }
    };
    const path = '/catalog/motoblochnoe-navesnoe-oborudovanie/adaptery-dlya-motoblokov';
    const id = 'adapter-k-motobloku-ara';
    const mockStore = configureStore();
    let store;
    it('should return null if no breadcrumbs exist', () => {
        const initialEntries = ['/test'];
        store = mockStore(initialState);
        render(
            <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                    <Breadcrumbs />
                </MemoryRouter>
            </Provider>
        );
        const element = screen.getByLabelText('breadcrumbs');
        expect(element).toBeEmptyDOMElement();
    });

    it('should return breadcrumbs', () => {
        store = mockStore(initialState);
        const initialEntries = [`${path}`];
        const { container } = render(
            <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                    <Breadcrumbs />
                </MemoryRouter>
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it('should return breadcrumbs with loader component', async () => {
        store = mockStore({ ...initialState, catalog: { detailed: { name: undefined, code: '', url: '' } } });
        const initialEntries = [`${path}/${id}`];
        const { container } = render(
            <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                    <Routes>
                        <Route path={`${path}/:id`} element={<Breadcrumbs />} />
                    </Routes>
                </MemoryRouter>
            </Provider>
        );
        const loader = await screen.findByRole('img');
        expect(loader).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('should return breadcrumbs with detailed product', async () => {
        store = mockStore(initialState);
        const initialEntries = [`${path}/${id}`];
        const { container } = render(
            <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                    <Routes>
                        <Route path={`${path}/:id`} element={<Breadcrumbs />} />
                    </Routes>
                </MemoryRouter>
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it('clicking on the button should open a menu of subcategories', () => {
        store = mockStore(initialState);
        const initialEntries = [`${path}`];
        render(
            <Provider store={store}>
                <MemoryRouter initialEntries={initialEntries}>
                    <Routes>
                        <Route path={path} element={<Breadcrumbs />} />
                    </Routes>
                </MemoryRouter>
            </Provider>
        );
        const button = screen.getByRole('presentation');
        const hiddenListOfSubcategories = screen.queryByRole('menu');
        expect(hiddenListOfSubcategories).toBeNull();
        fireEvent.click(button);
        const listOfSubcategories = screen.getByRole('menu');
        expect(listOfSubcategories).toBeInTheDocument();
        expect(listOfSubcategories).toMatchSnapshot();
    });
});
