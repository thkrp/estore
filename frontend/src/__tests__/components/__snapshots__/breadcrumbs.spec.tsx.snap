// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breadcrumbs clicking on the button should open a menu of subcategories 1`] = `
<div
  aria-label="sub-categories"
  class="sc-eDLKkx iDizGZ"
  role="menu"
>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/adaptery-dlya-motoblokov"
    role="menuitem"
  >
    Адаптеры для мотоблоков
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/borony-motoblochnye"
    role="menuitem"
  >
    Бороны
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/grabli-motoblochnye"
    role="menuitem"
  >
    Грабли мотоблочные
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/granulyatory-kormoizmelchiteli-zernodrobilki"
    role="menuitem"
  >
    Грануляторы, кормоизмельчители, зернодробилки
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/gruntozatsepy"
    role="menuitem"
  >
    Грунтозацепы
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/izmelchiteli-vetok-i-drovokoly"
    role="menuitem"
  >
    Измельчители веток и дровоколы
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/kartofelesazhalki"
    role="menuitem"
  >
    Картофелесажалки
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/kopalki-dlya-kartofelya-i-chesnoka-motoblochnye"
    role="menuitem"
  >
    Копалки для картофеля и чеснока мотоблочные
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/kosilki-k-motoblokam"
    role="menuitem"
  >
    Косилки к мотоблокам
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/mulchirovateli-dlya-motoblokov"
    role="menuitem"
  >
    Мульчирователи для мотоблоков
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/nabory-dlya-pereoborudovaniya"
    role="menuitem"
  >
    Наборы для переоборудования
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/okuchniki-pidgortachi-i-kultivatory"
    role="menuitem"
  >
    Окучники (пiдгортачi) и культиваторы
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/opryskivateli-motoblochnye"
    role="menuitem"
  >
    Опрыскиватели мотоблочные
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/ploskorezy-propolniki"
    role="menuitem"
  >
    Плоскорезы (пропольники)
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/plugi-motoblochnye"
    role="menuitem"
  >
    Плуги
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/pritsepy-dlya-motoblokov-i-mototraktorov"
    role="menuitem"
  >
    Прицепы для мотоблоков и мототракторов
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/razbrasyvateli-udobreniy-motoblochnye"
    role="menuitem"
  >
    Разбрасыватели удобрений мотоблочные
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/seyalki-motoblochnye"
    role="menuitem"
  >
    Сеялки мотоблочные
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/frezy-nozhi-dlya-pochvofrezy"
    role="menuitem"
  >
    Фрезы, ножи для почвофрезы
  </a>
  <a
    class="sc-jTQCzO gpVTaK"
    href="/catalog/motoblochnoe-navesnoe-oborudovanie/chesnochnye-sazhalki-dlya-motobloka"
    role="menuitem"
  >
    Чесночные сажалки
  </a>
</div>
`;

exports[`breadcrumbs should return breadcrumbs 1`] = `
<div>
  <div
    class="sc-beySPh dMUqEM"
  >
    <div
      aria-label="breadcrumbs"
      class="sc-guDLey ldurXD"
      role="navigation"
    >
      <div
        class="sc-dmyCSP HBqfv"
      >
        <a
          class="sc-jTQCzO gpVTaK"
          href="/catalog/motoblochnoe-navesnoe-oborudovanie"
        >
          Мотоблочное навесное оборудование
        </a>
        <div
          class="sc-hLQSwg cCIAir"
          role="presentation"
        />
      </div>
      <div
        class="sc-dmyCSP HBqfv"
      >
        Адаптеры для мотоблоков
      </div>
    </div>
  </div>
</div>
`;

exports[`breadcrumbs should return breadcrumbs with detailed product 1`] = `
<div>
  <div
    class="sc-beySPh dMUqEM"
  >
    <div
      aria-label="breadcrumbs"
      class="sc-guDLey ldurXD"
      role="navigation"
    >
      <div
        class="sc-dmyCSP HBqfv"
      >
        <a
          class="sc-jTQCzO gpVTaK"
          href="/catalog/motoblochnoe-navesnoe-oborudovanie"
        >
          Мотоблочное навесное оборудование
        </a>
        <div
          class="sc-hLQSwg cCIAir"
          role="presentation"
        />
      </div>
      <div
        class="sc-dmyCSP HBqfv"
      >
        <a
          class="sc-jTQCzO gpVTaK"
          href="/catalog/motoblochnoe-navesnoe-oborudovanie/adaptery-dlya-motoblokov"
        >
          Адаптеры для мотоблоков
        </a>
      </div>
      <div
        class="sc-dmyCSP HBqfv"
      >
        Адаптер к мотоблоку "АРА"
      </div>
    </div>
  </div>
</div>
`;

exports[`breadcrumbs should return breadcrumbs with loader component 1`] = `
<div>
  <div
    class="sc-beySPh dMUqEM"
  >
    <div
      aria-label="breadcrumbs"
      class="sc-guDLey ldurXD"
      role="navigation"
    >
      <div
        class="sc-dmyCSP HBqfv"
      >
        <a
          class="sc-jTQCzO gpVTaK"
          href="/catalog/motoblochnoe-navesnoe-oborudovanie"
        >
          Мотоблочное навесное оборудование
        </a>
        <div
          class="sc-hLQSwg cCIAir"
          role="presentation"
        />
      </div>
      <div
        class="sc-dmyCSP HBqfv"
      >
        <a
          class="sc-jTQCzO gpVTaK"
          href="/catalog/motoblochnoe-navesnoe-oborudovanie/adaptery-dlya-motoblokov"
        >
          Адаптеры для мотоблоков
        </a>
      </div>
      <div
        class="sc-dmyCSP HBqfv"
      >
        <svg
          height="15px"
          role="img"
        >
          horizontal-loader.svg
        </svg>
      </div>
    </div>
  </div>
</div>
`;
